### PR TITLE
Add generics to ModelManagerInterface

### PR DESCRIPTION
**Pull request by core23**
## Subject

Add some more generics to our project.

I am targeting this branch, because this is BC.

## Changelog

```markdown
### Added
- Add generics to `ModelManagerInterface`
```

### CODE REVIEW

**Comment by phansys** [comment]
IMO, these changes seem pedantic, so I think the "Changelog" section should be omitted.

**Comment by greg0ire** [comment]
They do affect the end user though, don't they? In the CI and in the IDE.

**Comment by core23** [comment]
> IMO, these changes seem pedantic, so I think the "Changelog" section should be omitted.

We marked these changes as `minor` on the datagrid bundle: https://github.com/sonata-project/SonataDatagridBundle/pull/226

File: src/Model/ModelManagerInterface.php [hunk 1]
```diff
@@ -35,35 +37,49 @@ public function getNewFieldDescriptionInstance($class, $name, array $options = [
...
      */
     public function delete($object);
 
     /**
      * @param string $class
      *
-     * @return array all objects matching the criteria
+     * @return object[] all objects matching the criteria
+     *
+     * @psalm-template C
```
Comment (greg0ire) [inline src/Model/ModelManagerInterface.php:69]: Is `C` supposed to be a subclass of `T` here? If yes, shouldn't we have the following?
```suggestion
     * @psalm-template C of T
```

Comment (core23) [reply src/Model/ModelManagerInterface.php:69]: I'm not sure if it must be a subclass or not

File: src/Model/ModelManagerInterface.php [hunk 2]
```diff
@@ -21,6 +21,8 @@
 
 /**
  * A model manager is a bridge between the model classes and the admin functionality.
+ *
+ * @psalm-template T
```
Comment (jordisala1991) [inline src/Model/ModelManagerInterface.php:25]: can you use phpstan instead? thats what I am using on master

Comment (jordisala1991) [reply src/Model/ModelManagerInterface.php:25]: Also you should look like at the recent merged pr on master, all of this is added there, the work should be backported

Comment (core23) [reply src/Model/ModelManagerInterface.php:25]: I'll do a backport

Comment (core23) [reply src/Model/ModelManagerInterface.php:25]: > can you use phpstan instead? thats what I am using on master

phpstan does not provide advanced annotations like `class-string`

Comment (jordisala1991) [reply src/Model/ModelManagerInterface.php]: 👍 I am doing a Narrow the Api on master branch, and in general it is a bc break. Some parts could be bc compat but spliting the already hard work makes it super hard. So if someone wants to backport the bc compat parts I am happy with it, less conflicts

**Comment by VincentLanglet** [comment]
What do you think about adding something in https://github.com/sonata-project/dev-kit/blob/master/templates/project/CONTRIBUTING.md about the static analysis annotation we prefer to use (@psalm vs @phpstan) ?

**Comment by greg0ire** [comment]
Might be a good idea, we have many such unwritten rules and it would be great to write them down somewhere. Not sure if CONTRIBUTING.md is the right place, if it gets too cluttered in the long run we should consider something like a separate style guide.

**Review by jordisala1991 — APPROVED** [review]

**Comment by jordisala1991** [comment]
Thank you @core23

**Pull request merged**